system-x/services/kudu/src/main/java/software/tnb/kudu/resource/openshift/OpenshiftKudu.java

```diff
@@ -0,0 +1,379 @@
+package software.tnb.kudu.resource.openshift;
+
+import software.tnb.common.config.OpenshiftConfiguration;
+import software.tnb.common.deployment.OpenshiftDeployable;
+import software.tnb.common.deployment.WithExternalHostname;
+import software.tnb.common.openshift.OpenshiftClient;
+import software.tnb.common.utils.WaitUtils;
+import software.tnb.kudu.service.Kudu;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.auto.service.AutoService;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import io.fabric8.kubernetes.api.model.ContainerPort;
+import io.fabric8.kubernetes.api.model.ContainerPortBuilder;
+import io.fabric8.kubernetes.api.model.EnvVar;
+import io.fabric8.kubernetes.api.model.EnvVarSource;
+import io.fabric8.kubernetes.api.model.IntOrString;
+import io.fabric8.kubernetes.api.model.ObjectFieldSelector;
+import io.fabric8.kubernetes.api.model.Pod;
+import io.fabric8.kubernetes.api.model.Probe;
+import io.fabric8.kubernetes.api.model.ProbeBuilder;
+import io.fabric8.kubernetes.api.model.ServiceBuilder;
+import io.fabric8.kubernetes.api.model.ServicePortBuilder;
+import io.fabric8.kubernetes.api.model.ServiceSpecBuilder;
+import io.fabric8.kubernetes.api.model.VolumeBuilder;
+import io.fabric8.kubernetes.api.model.VolumeMountBuilder;
+import io.fabric8.kubernetes.api.model.apps.DeploymentBuilder;
+import io.fabric8.kubernetes.api.model.apps.StatefulSetBuilder;
+import io.fabric8.openshift.api.model.RouteBuilder;
+
+@AutoService(Kudu.class)
+public class OpenshiftKudu extends Kudu implements OpenshiftDeployable, WithExternalHostname {
+
+    private static final Logger LOG = LoggerFactory.getLogger(OpenshiftKudu.class);
+
+    final List<String> hostNames = List.of(MASTER_PREFIX, TSERVER_PREFIX);
+    private static final String KUDU_CLIENT = "kudu-client";
+
+    @Override
+    public void openResources() {
+        client = new RemoteClient(getMastersUsingRpc(), getClientPodName());
+        LOG.debug("master leader {}", client.findLeaderMasterServer());
+        createUIRoute();
+    }
+
+    private String getClientPodName() {
+        return OpenshiftClient.get().pods().withLabel(OpenshiftConfiguration.openshiftDeploymentLabel(), KUDU_CLIENT)
+            .list().getItems().get(0).getMetadata().getName();
+    }
+
+    @Override
+    public void closeResources() {
+        if (client != null) {
+            try {
+                client.close();
+                client = null;
+            } catch (IOException e) {
+                throw new RuntimeException(e);
```

Comment:
isn't `LOG.warn` enough here? why fail the test when it is done, but couldn't close the client

Reply:
yes, code auto-completion....